src/core/connection_pool.c

```diff
@@ -0,0 +1,490 @@
+    create a pool of connections that are spread across RSS cores.
+
+--*/
+
+#include "precomp.h"
+
+#ifdef QUIC_CLOG
+#include "connection_pool.c.clog.h"
+#endif
+
+#define MAX_CONNECTION_POOL_RETRY_MULTIPLIER 2
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+QuicConnPoolGetStartingLocalAddress(
+    _In_ QUIC_ADDR* RemoteAddress,
+    _Out_ QUIC_ADDR* LocalAddress
+    )
+{
+    CXPLAT_SOCKET* Socket = NULL;
+    CXPLAT_UDP_CONFIG UdpConfig = { .RemoteAddress = RemoteAddress, };
+    QUIC_STATUS Status =
+        CxPlatSocketCreateUdp(MsQuicLib.Datapath, &UdpConfig, &Socket);
+    if (QUIC_FAILED(Status)) {
+        goto Error;
+    }
+
+    CxPlatSocketGetLocalAddress(Socket, LocalAddress);
+
+Error:
+    if (Socket != NULL) {
+        CxPlatSocketDelete(Socket);
+    }
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+QuicConnPoolGetInterfaceIndexForLocalAddress(
+    _In_ QUIC_ADDR* LocalAddress,
+    _Out_ uint32_t* InterfaceIndex
+    )
+{
+    CXPLAT_ADAPTER_ADDRESS* Addresses = NULL;
+    uint32_t AddressesCount;
+
+    *InterfaceIndex = 0;
+
+    QUIC_STATUS Status =
+        CxPlatDataPathGetLocalAddresses(
+            MsQuicLib.Datapath,
+            &Addresses,
+            &AddressesCount);
+    if (QUIC_FAILED(Status)) {
+        goto Error;
+    }
+
+    for (uint32_t i = 0; i < AddressesCount; i++) {
+        if (QuicAddrCompareIp(LocalAddress, &Addresses[i].Address)) {
+            *InterfaceIndex = Addresses[i].InterfaceIndex;
+            break;
+        }
+    }
+
+    if (InterfaceIndex == 0) {
+        Status = QUIC_STATUS_NOT_FOUND;
+        QuicTraceLogError(
+            ConnPoolLocalAddressNotFound,
+            "[conp] Failed to find local address, 0x%x",
+            Status);
+        goto Error;
+    }
+
+Error:
+
+    if (Addresses != NULL) {
+        CXPLAT_FREE(Addresses, QUIC_POOL_DATAPATH_ADDRESSES);
+    }
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+QUIC_API
+MsQuicConnectionPoolCreate(
+    _In_ QUIC_CONNECTION_POOL_CONFIG* Config,
+    _Out_writes_(Config->NumberOfConnections)
+        HQUIC* ConnectionPool
+    )
+{
+    QUIC_CONNECTION** Connections = (QUIC_CONNECTION**)ConnectionPool;
+    CXPLAT_RSS_CONFIG* RssConfig = NULL;
+    uint32_t* RssProcessors = NULL;
+    uint32_t* ConnectionCounts = NULL;
+    CXPLAT_TOEPLITZ_HASH ToeplitzHash;
+    QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
+    uint32_t CreatedConnections = 0;
+
+    if (Config == NULL || Config->NumberOfConnections == 0 || Config->Handler == NULL ||
+        Config->ServerName == NULL || Config->ServerPort == 0 || ConnectionPool == NULL ||
+        (Config->CibirIds != NULL && Config->CibirIdLength == 0)) {
+        Status = QUIC_STATUS_INVALID_PARAMETER;
+        QuicTraceLogError(
+            ConnPoolInvalidParam,
+            "[conp] Invalid parameter, 0x%x",
+            Status);
+        goto Error;
```

Comment:
Just return instead? Because otherwise you have an unmatched entry/exit API log...